test/unit/signal_wait_until.c

```diff
@@ -0,0 +1,105 @@
+/*
+ *  Copyright (c) 2021 Intel Corporation. All rights reserved.
+ *  This software is available to you under the BSD license below:
+ *
+ *      Redistribution and use in source and binary forms, with or
+ *      without modification, are permitted provided that the following
+ *      conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+/* 
+ * Validate signal_wait_until operation using blocking put_signal
+*/
+
+#include <stdio.h>
+#include <shmem.h>
+#include <string.h>
+
+#define MSG_SZ 10
+
+int main(int argc, char *argv[])
+{
+    long source[MSG_SZ];
+    long *target;
+    int me, npes, i;
+    int errors = 0;
+
+    static uint64_t sig_addr = 0;
+
+    shmem_init();
+
+    me = shmem_my_pe();
+    npes = shmem_n_pes();
+
+    for (i = 0; i < MSG_SZ; i++)
+        source[i] = i;
+
+    target = (long *) shmem_calloc(MSG_SZ, sizeof(long));
+    if (!target) {
+        fprintf(stderr, "Failed to allocate target pointer\n");
+        shmem_global_exit(1);
+    }
+
+    shmem_barrier_all();
+    for (i = 0; i < npes; i++) {
+        shmem_long_put_signal(target, source, MSG_SZ, &sig_addr, 1, SHMEM_SIGNAL_ADD, i);
+    }
+
+    shmem_signal_wait_until(&sig_addr, SHMEM_CMP_EQ, npes);
+
+    if (sig_addr != npes)
+        errors++;
+
+    for (i = 0; i < MSG_SZ; i++) {
+        if (target[i] != source[i]) {
+            fprintf(stderr, "%10d: target[%d] = %ld not matching %ld with SHMEM_SIGNAL_ADD\n",
+                    me, i, target[i], source[i]);
+            errors++;
+        }
+    }
+
+    for (i = 0; i < MSG_SZ; i++)
+        target[i] = 0;
+
+    shmem_barrier_all();
+
+    for (i = 0; i < npes; i++) {
+        shmem_long_put_signal(target, source, MSG_SZ, &sig_addr, 1, SHMEM_SIGNAL_SET, i);
+    }
+
+    shmem_signal_wait_until(&sig_addr, SHMEM_CMP_EQ, 1);
```

Comment:
In equality comparison, its not required to check the return value. In other comparisons, we might have to check the return value. Also, do we have to check the value in such cases with the return value from `shmem_signal_fetch`?

Reply:
Ooo, good point.

I think we should check the return value in this test to make sure it's what we expect (1 in this case, `npes` above).  Does that sound good?

I think I understand what you mean about "other comparison", but I'd say the _application_ determines whether it's required to check the return value.  Is that what you mean?  I think the same probably applies to `shmem_signal_fetch`.

I'm not sure we have great testing coverage over the different comparisons for all the varieties of `shmem_wait` and `shmem_test`.  I don't see any coverage for `SHMEM_CMP_LE` and `SHMEM_CMP_LT` for example, but it makes sense that they wouldn't naturally be used very often.  Maybe I could create an issue for that?

Reply:
Yes, makes sense. 

I meant to say, for equality case, even if we dont check the return value, we know what `sig_addr` has now as we already just did the equality test. But, for any other comparison, the application might have to check for the return value.. so, we may want to add one such test. I am not really sure when would be that case providing we can only do `SIGNAL_ADD` and `SIGNAL_SET` .. and we will always know what the `sig_addr` is supposed to have at any point. Not quite sure why we added the return value here... 

It is also getting confusing that different `wait_*` routines return different value semantics. Should we unify these? Thoughts?

Reply:
Ah, that makes sense - I'll see if I can change the test a bit to cover the non-equality comparisons as well.

Reply:
Thanks for the changes. The test looks good to me now.